File: typescipt/src/language/rootNodeList.ts


export class RootNodeList extends Array<IRootNode> {
   private readonly Array<IRootNode> values = collection<IRootNode>(): new;

   public number Count => values.Count;

   public getEnumerator(): IEnumerator<IRootNode> {
     return values.GetEnumerator();
   }

   IEnumerator IEnumerable.GetEnumerator() {
     return GetEnumerator();
   }

   public add(rootNode: IRootNode): void {
     values.Add(rootNode);
   }

   internal containsEnum(enumName: string): boolean {
     return values
       .OfType<EnumDefinition>()
       .Any(definition => definition.Name.Value == enumName);
   }

   public getNode(name: string): IRootNode {
     return values
       .FirstOrDefault(definition => definition.NodeName == name);
   }

   public contains(name: string): boolean {
     return values
       .Any(definition => definition.NodeName == name);
   }

   public getFunction(name: string): Function {
     return values
       .OfType<Function>()
       .FirstOrDefault(function => function.Name.Value == name);
   }

   public getTable(name: string): Table {
     return values
       .OfType<Table>()
       .FirstOrDefault(function => function.Name.Value == name);
   }

   public getCustomType(name: string): TypeDefinition {
     return values
       .OfType<TypeDefinition>()
       .FirstOrDefault(function => function.Name.Value == name);
   }

   public getSingleFunction(): Function {
     return values
       .OfType<Function>()
       .SingleOrDefault();
   }

   public getScenarios(): Array<Scenario> {
     return values.OfType<Scenario>();
   }

   public getEnum(name: string): EnumDefinition {
     return values
       .OfType<EnumDefinition>()
       .FirstOrDefault(enumDefinition => enumDefinition.Name.Value == name);
   }

   public addIfNew(node: IRootNode): void {
     if (!values.contains(node)) values.Add(node);
   }

   public first(): INode {
     return values.FirstOrDefault();
   }

   public getType(name: string): TypeWithMembers {
     let node = GetNode(name);
     return node switch {
       Table table => new TableType(name, table),
       Function function => new FunctionType(name, function),
       EnumDefinition enumDefinition => new EnumType(name, enumDefinition),
       TypeDefinition typeDefinition => new CustomType(name, typeDefinition),
       _ => null
     };
   }
}

File: src/Lexy.Compiler/Language/RootNodeList.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Lexy.Compiler.Language.Enums;
using Lexy.Compiler.Language.Functions;
using Lexy.Compiler.Language.Scenarios;
using Lexy.Compiler.Language.Types;
using Lexy.Compiler.Language.VariableTypes;
using Table = Lexy.Compiler.Language.Tables.Table;

namespace Lexy.Compiler.Language;

public class RootNodeList : IRootNodeList
{
    private readonly IList<IRootNode> values;

    public RootNodeList(IEnumerable<IRootNode> values = null)
    {
        this.values = values != null ? values.ToList() : new List<IRootNode>();
    }

    public IEnumerator<IRootNode> GetEnumerator()
    {
        return values.GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    public void Add(IRootNode rootNode)
    {
        values.Add(rootNode);
    }

    internal bool ContainsEnum(string enumName)
    {
        return values
            .OfType<EnumDefinition>()
            .Any(definition => definition.Name.Value == enumName);
    }

    public IRootNode GetNode(string name)
    {
        return values
            .FirstOrDefault(definition => definition.NodeName == name);
    }

    public bool Contains(string name)
    {
        return values
            .Any(definition => definition.NodeName == name);
    }

    public Function GetFunction(string name)
    {
        return values
            .OfType<Function>()
            .FirstOrDefault(function => function.Name.Value == name);
    }

    public Table GetTable(string name)
    {
        return values
            .OfType<Table>()
            .FirstOrDefault(function => function.Name.Value == name);
    }

    public TypeDefinition GetCustomType(string name)
    {
        return values
            .OfType<TypeDefinition>()
            .FirstOrDefault(function => function.Name.Value == name);
    }

    public IEnumerable<Scenario> GetScenarios()
    {
        return values.OfType<Scenario>();
    }

    public EnumDefinition GetEnum(string name)
    {
        return values
            .OfType<EnumDefinition>()
            .FirstOrDefault(enumDefinition => enumDefinition.Name.Value == name);
    }

    public void AddIfNew(IRootNode node)
    {
        if (!values.Contains(node)) values.Add(node);
    }

    public INode First()
    {
        return values.FirstOrDefault();
    }

    public TypeWithMembers GetType(string name)
    {
        var node = GetNode(name);
        return node switch
        {
            Table table => new TableType(name, table),
            Function function => new FunctionType(name, function),
            EnumDefinition enumDefinition => new EnumType(name, enumDefinition),
            TypeDefinition typeDefinition => new CustomType(name, typeDefinition),
            _ => null
        };
    }
}